### PR TITLE
Implement FromGlibPtrNone and FromGlibPtrFull for Value

### PR DESCRIPTION
**Pull request by sdroege**
There are some functions that need that in GStreamer.

### CODE REVIEW

**Comment by EPashkin** [comment]
Maybe better add 2 other options but IMHO it can wait (don't like add unused functions)
@GuillaumeGomez 👍 for this

**Comment by sdroege** [comment]
I can also add the other trait impls, might be good for consistency

File: src/value.rs [hunk 1]
```diff
@@ -242,6 +242,25 @@ impl<'a> ToGlibPtr<'a, *mut gobject_ffi::GValue> for &'a [&'a ToValue] {
...
+        ret
+    }
+}
+
+impl FromGlibPtrFull<*mut gobject_ffi::GValue> for Value {
+    unsafe fn from_glib_full(ptr: *mut gobject_ffi::GValue) -> Self {
+        let mut ret = Value::uninitialized();
+        ptr::swap(&mut ret.0, ptr);
+        glib_ffi::g_free(ptr as *mut c_void);
+
```
Comment (GuillaumeGomez) [inline src/value.rs]: I don't think this line is necessary. ;)

Comment (sdroege) [inline src/value.rs]: True :)

**Comment by sdroege** [comment]
So should I implement impls for the other traits (which?) or we just wait until someone needs it? :)

**Comment by EPashkin** [comment]
IMHO it good as is.

**Comment by antoyo** [comment]
I also implemented these traits in [my pull request](https://github.com/gtk-rs/glib/pull/173/files#diff-dd26c2766e65aca43e4228e64ebd0cf8R220).
I'm not sure which way is better but mine doesn't involve a copy (not sure if it is correct, thought).

By the way, are some implementions missing in this pull request?
I've got 4 implementations in mine.

**Comment by sdroege** [comment]
@antoyo Sorry, I should've checked before. Didn't expect another pull request to do this as it's rather uncommon for C code to give you GValue pointers.

See my review comments on your code. I could add the `impl FromGlibPtrNone<*mut gobject_ffi::GValue> for Value` as that one is technically correct, but the `impl FromGlibPtrFull<*const gobject_ffi::GValue> for Value` is not. Anyway, see comments there :)

Should we finish your pull request and get that one merged because you were first, or just merge this one?

**Comment by sdroege** [comment]
I added `impl FromGlibPtrNone<*mut gobject_ffi::GValue> for Value` too here. I guess let's get in my pull request for now, and @antoyo 's then only for the GClosure stuff (thanks btw, I will also need that soonish)

**Comment by sdroege** [comment]
@antoyo in your case you probably also want to prevent the copying to happen (you just want to borrow the values from the arguments to the `marshal` function), for which it would make sense to add a lifetime parameter to FromGlibPtrNone (and maybe Full too?), and then implement it for `&'a Value` too

**Comment by antoyo** [comment]
I don't have any problem on basing my pull request on this work (thanks for you work too ;) ).

I read about memory management in glib and I wonder if `FromGlibPtrNone` from glib is actually correct.
From what I understand, when there's no ownership transfer, it is similar to `&T` in Rust, so I believe `from_glib_none` should actually return a reference.
But I can see in the documentation that `FromGlibPtrNone<*const c_char> for String` is implemented and that copies the `String`.

I don't understand yet how all these traits work.
Do I need `FromGlibPtrBorrow` to get a reference to a value that I get with transfert none?
If yes, I would need the implementation of this trait in this pull request.

**Comment by sdroege** [comment]
Basically what we would need for borrowing is `impl<'a> FromGlibPtrNone<*const gobject_ffi::GValue> for &'a Value` for example. Currently all implementations of `FromGlibPtrNone` are doing a copy (or another `ref()`), which is also correct. "transfer none" in GLib just means that you, as the caller of that function, don't own the return value. So taking a copy is a valid thing to do in that case, while with "transfer full" we could just take the return value without copying (but often can't because of different allocators, etc).

**Comment by sdroege** [comment]
See also https://github.com/gtk-rs/glib/pull/162 for a similar case

**Comment by sdroege** [comment]
@GuillaumeGomez Anything blocking this from being merged?

**Comment by GuillaumeGomez** [comment]
No, we're all good. Thanks!

**Comment by sdroege** [comment]
Great, thanks :)

**Pull request merged**